code/lib/instrumenter/src/instrumenter.ts

```diff
@@ -497,13 +497,20 @@ export class Instrumenter {
           this.setState(call.storyId, { cursor: 0, ancestors: [...ancestors, call.id] });
           const restore = () => this.setState(call.storyId, { cursor, ancestors });
 
-          // Invoke the actual callback function.
-          const res = arg(...args);
-
-          // Reset cursor and ancestors to their original values before we entered the callback.
-          if (res instanceof Promise) return res.then(restore, restore);
-          restore();
-          return res;
+          let restored = false;
+          try {
+            // Invoke the actual callback function.
+            const res = arg(...args);
+
+            // Reset cursor and ancestors to their original values before we entered the callback.
```

Comment:
I'd hoist up the two comments to before the `try`, because there's now multiple places where the reset might happen. E.g. 
```js
// Invoke the actual callback function, then reset cursor and ancestors to their original values before we entered the callback.
try {
  ...
```
Then I'd remove the empty newline (504).